### PR TITLE
471 environmental measures

### LINKED ISSUE #471: environmental measures: new object and codelist
# Background

The new Environment Module included the data point Environmental Measures:

"Disclose the measures adopted by the project to mitigate and/or remedy the environmental impact (E.g. [free text justification/explaining the measures adopted] and document)."

This has been mapped to `document`. A new draft of the data point suggests updating it to the following:

Environmental measures: Identify the measures adopted by the project to mitigate and/or remedy the environmental impact, disclosing the corresponding document that describes the project’s environmental measures. This can include, without limitation, the following:

* waste management
* disposal of construction by-products
* pollution reduction (including noise and dust)
* environmentally responsible sourcing of materials
* environmentally responsible use of land, water and air
* water contamination management
* use of recycled and reused materials
* others (explain)

Other similar data points, e.g. Conservation measures, Climate measures, have been modelled as new objects of the format:

```json
{
  "type": [array from codelist],
  "description": "free-text"
}
```
We could leave the modelling as it is and add a line to the mapping to use `document.description` to record the list items. Or we could create a new object and codelist in a similar manner to https://github.com/open-contracting/infrastructure/issues/402.

@jpmckinney do you have any thoughts? Leaving it as it is and using the `document.description` field would be simpiler but adding the new object would be more consistent.

### PR DESCRIPTION
**Pull request by odscjen**
**Related issues**

closes #471 

**Description**

**Merge checklist**

- [x] [Log your changes](https://ocds-standard-development-handbook.readthedocs.io/en/latest/standard/contributing.html#logging-changes)

If there are changes to `project-schema.json` or `project-package-schema.json`:

- [x] Update the examples:
  - [ ] `docs/examples/example.json`
  - [ ] `docs/examples/blank.json`
- [x] Run `./manage.py pre-commit` to update `docs/_static/i8n.csv`

If you added a new definition to the schema, run `./manage.py pre-commit`.

If you added a new codelist:

  - [x] Add an entry to `docs/reference/codelists.md`

### CODE REVIEW

File: schema/project-level/codelists/environmentalMeasure.csv [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Code,Title,Description
+wasteManagement,Waste Management,"Measures that avoid, reduce, recycle and reuse waste generated as a direct result of the project. When waste is produced, it shall be treated, destroyed or disposed of in a safe and environmentally sound manner."
+constructionWasteDisposal,Disposal of construction by-products,"Measures that avoid, reduce, recycle and reuse construction by-products generated as a direct result of the project. When construction by-products are produced, they shall be treated, destroyed or disposed of in a safe and environmentally sound manner."
+materialSourcing,Environmentally responsible sourcing of materials,"Measures that ensure all project materials are sourced in an environmentally sustainable manner. For example, Leadership in Energy and Environmental Design's (LEED) for building materials or the Forest Stewardship Council's (FSC) for wood."
+useNaturalResources,"Environmentally responsible use of land, water and air","Measures that ensure the sustainable and environmentally responsible use of land, water and air by the project."
```
Comment (neelima-j) [inline schema/project-level/codelists/environmentalMeasure.csv]:
```suggestion
naturalResourceUse,"Environmentally responsible use of land, water and air","Measures that ensure the sustainable and environmentally responsible use of land, water and air by the project."
```
as per https://github.com/open-contracting/infrastructure/issues/471#issuecomment-1936844236

File: schema/project-level/codelists/environmentalMeasure.csv [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Code,Title,Description
+wasteManagement,Waste Management,"Measures that avoid, reduce, recycle and reuse waste generated as a direct result of the project. When waste is produced, it shall be treated, destroyed or disposed of in a safe and environmentally sound manner."
```
Comment (neelima-j) [inline schema/project-level/codelists/environmentalMeasure.csv]:
```suggestion
wasteManagement,Waste Management,"Measures that avoid, reduce, recycle and reuse waste generated as a direct result of the project. When waste is produced, it ought to be treated, destroyed or disposed of in a safe and environmentally sound manner."
```
as per https://github.com/open-contracting/infrastructure/issues/471#issuecomment-1936844236

File: schema/project-level/codelists/environmentalMeasure.csv [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+constructionWasteDisposal,Disposal of construction by-products,"Measures that avoid, reduce, recycle and reuse construction by-products generated as a direct result of the project. When construction by-products are produced, they shall be treated, destroyed or disposed of in a safe and environmentally sound manner."
```
Comment (neelima-j) [inline schema/project-level/codelists/environmentalMeasure.csv]:
```suggestion
constructionWasteDisposal,Disposal of construction by-products,"Measures that avoid, reduce, recycle and reuse construction by-products generated as a direct result of the project. When construction by-products are produced, they ought to be treated, destroyed or disposed of in a safe and environmentally sound manner."
```
as per https://github.com/open-contracting/infrastructure/issues/471#issuecomment-1936844236

File: schema/project-level/codelists/environmentalMeasure.csv [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+materialSourcing,Environmentally responsible sourcing of materials,"Measures that ensure all project materials are sourced in an environmentally sustainable manner. For example, Leadership in Energy and Environmental Design's (LEED) for building materials or the Forest Stewardship Council's (FSC) for wood."
```
Comment (neelima-j) [inline schema/project-level/codelists/environmentalMeasure.csv]:
```suggestion
materialSourcing,Environmentally responsible sourcing of materials,"Measures that ensure all project materials are sourced in an environmentally sustainable manner. For example, Leadership in Energy and Environmental Design (LEED) standards for building materials or Forest Stewardship Council (FSC) standards for wood."
```
Added the word 'standard'.

File: mapping/sustainability.yaml [hunk 5]
```diff
@@ -355,13 +355,31 @@
...
+        - water contamination management
+        - others (explain)
     mapping: |-
         Project level:
 
-        [Add a project document](../common.md#add-a-project-document), set its [`.documentType`](project-schema.json,/definitions/Document,documentType) to 'environmentalMeasures' and map the justification and explanation to its [`.description`](project-schema.json,/definitions/Document,description).
+        1. For each measure, add an [`EnvironmentalMeasure`](../../reference/schema.md#environmentalmeasure) object to the [`environment.environmentalMeasures`](project-schema.json,,environment/environmentalMeasures) array, set its [`.type`](project-schema.json,/definitions/EnvironmentalMeasure,type) according to the open [environmentalMeasure](../../reference/codelists.md#environmentalmeasure) codelist and map any further details to its [`.description`](project-schema.json,/definitions/EnvironmentalMeasure,description).
+
+        2. If there is an accompanying document, [add a project document](../common.md#add-a-project-document) and set its [`.documentType`](project-schema.json,/definitions/Document,documentType) to 'environmentalMeasures'.
```
Comment (neelima-j) [inline mapping/sustainability.yaml:372]: I think the environmentalMeasures documentType hasn't been added.

File: docs/reference/changelog.md [hunk 6]
```diff
@@ -89,6 +90,7 @@
     * debtRepaymentPriority
     * financingArrangementType
     * financingPartyType
+  * [#476](https://github.com/open-contracting/infrastructure/pull/476) - environmentalMeasures
```
Comment (neelima-j) [inline docs/reference/changelog.md]:
```suggestion
  * [#476](https://github.com/open-contracting/infrastructure/pull/476) - environmentalMeasure
```
The codelist is environmentalMeasure. Similarly, codelist in line 74 should now be climateMeasure.

File: docs/reference/changelog.md [hunk 7]
```diff
@@ -117,7 +119,7 @@
       * 'minutes.consultationMeeting'
       * 'minutes.lobbyingMeeting'
     * [#442](https://github.com/open-contracting/infrastructure/pull/442) - 'abatementCostMethodology'
-    * [#457](https://github.com/open-contracting/infrastructure/pull/457) - 'climateMeasures'
+    * [#457](https://github.com/open-contracting/infrastructure/pull/457) - 'climateMeasure'
```
Comment (neelima-j) [inline docs/reference/changelog.md]:
```suggestion
    * [#457](https://github.com/open-contracting/infrastructure/pull/457) - 'climateMeasures'
```
This code is still in the plural in documentType- climateMeasures.

**Review by neelima-j — CHANGES_REQUESTED** [review]
- 'environmentalMeasures' to be added to documentType 
 - Other changes suggested to match issue discussion.

File: docs/reference/changelog.md (hunk 7 above)
Comment (odscjen) [reply docs/reference/changelog.md]: the documentType I think should still be plural as the document could be detailing multiple measures

File: mapping/sustainability.yaml (hunk 5 above)
Comment (odscjen) [reply mapping/sustainability.yaml:372]: good catch! that should have been added as part of a previous PR that added a LOT of new codes to documentType

**Comment by odscjen** [comment]
@jpmckinney this should be the final PR for changes to 0.9.4. Once this is merged we can merge 0.9-dev into 0.9 and then make a start on the Spanish translation of the new bits.

File: docs/reference/changelog.md [hunk 8]
```diff
@@ -62,6 +62,7 @@
   * [#441](https://github.com/open-contracting/infrastructure/pull/441) - `environment.impactCategories`
   * [#455](https://github.com/open-contracting/infrastructure/pull/455) - `environment.inProtectedArea`
   * [#428](https://github.com/open-contracting/infrastructure/pull/428) - `lobbyingMeetings`
+  * [#476](https://github.com/open-contracting/infrastructure/pull/476) - `environment.environmentalMeasures`
```
Comment (jpmckinney) [inline docs/reference/changelog.md:65]: Move this into alphabetical order per field name (not issue number).

File: schema/project-level/codelists/documentType.csv [hunk 9]
```diff
@@ -93,3 +93,4 @@ consultationMeetingInvitation,Consultation meeting invitation,A notice about an
 materialTestResults,Material test results,Documentation of results of tests conducted on construction materials.,OC4IDS
 climateMeasures,Climate measures,Documentation of the measures adopted by the project to adapt to or mitigate against climate change.,OC4IDS
 maintenancePlan,Maintenance plan,Documentation that describes work to prevent the breakdown or malfunctioning of an asset.,OC4IDS
+environmentalMeasures,Environmental measures,Documentation of the measures adopted by the project to mitigate or remedy environmental impacts.,OC4IDS
```
Comment (jpmckinney) [inline schema/project-level/codelists/documentType.csv]: Move this and climateMeasures under conservationMeasures (and order them alphabetically).

File: schema/project-level/project-schema.json [hunk 10]
```diff
@@ -460,6 +460,16 @@
           "uniqueItems": true,
           "minItems": 1
         },
+        "environmentalMeasures": {
```
Comment (jpmckinney) [inline schema/project-level/project-schema.json:473]: As with the codelist, move this and climateMeasures under conservationMeasures (and order them alphabetically).

**Review by jpmckinney — COMMENTED** [review]
Just a few suggestions to make the codelist and schema easier to read.

**Comment by odscjen** [comment]
Thanks @jpmckinney I've made those updates, is this ok to merge now?

**Review by jpmckinney — APPROVED** [review]

**Pull request merged**